infra/Pos/Infra/Diffusion/Types.hs

```diff
@@ -36,6 +47,12 @@ data Diffusion m = Diffusion
                          -> HeaderHash
                          -> [HeaderHash]
                          -> m (OldestFirst [] Block)
+    , streamBlocks       :: forall t .
+                            NodeId
+                         -> HeaderHash
+                         -> [HeaderHash]
+                         -> ((Word32, Maybe Gauge, TBQueue StreamEntry) -> m t)
+                         -> m (Maybe t)
```

Comment:
This type doesn't look ideal. Why should a `Gauge` appear here? The particular `Diffusion` implementation should deal with that.

Reply:
`Gauge` is the Write Queue performance counter which is incremented in the network conversation thread and decremented in the thread which writes blocks to store.

Reply:
Does it need to be passed to the caller, though? Couldn't we have the gauge be local to the full diffusion layer? How about this: replace the `TBQueue StreamEntry` with `m StreamEntry`, and then for the full diffusion implementation, it's

```
dequeue :: IO StreamEntry
dequeue = do
  next <- atomically $ readTBQueue queue
  _ <- updateGauge wgqM next
  pure next
```

Reply:
The `((Word32, Maybe Gauge, TBQueue StreamEntry) -> m t)` callback function exists because `streamBlocks` can't call `handleBlocks` and the `Maybe Gauge` argument exists because `streamProcessBlocks` can't access Difussion EKG counters. It is a mess, but that mess was caused by the existing design.

I may be misunderstanding your dequeue implementation but it does not do what the `loop` function in `streamProcessBlocks` does.

Reply:
> I may be misunderstanding your dequeue implementation but it does not do what the loop function in streamProcessBlocks does.

Indeed it doesn't, and shouldn't. `streamProcessBlocks` takes a `TBQueue` right now, but instead it would take this `IO StreamEntry` that does the gauge updating along with dequeueing from the `TBQueue`.